### DISCUSSION TITLE
Cannot join new auth/proxy servers to cluster

### DISCUSSION THREAD

**Discussion by Ruchira-R** [gravitational/teleport#19973] [Q&A]
Expected behavior: Should connect to the cluster

Current behavior: Proxy Failed to establish connection to cluster
Proxy failed to establish connection to cluster: connection error: desc = "transport: Error while dialing failed to dial: dial tcp lb-ip:3025: i/o timeout". pid:11409.1 service/connect.go:119

Bug details:

Teleport version : Teleport v11.1.4 git:v11.1.4-0-ge4ac5f6 go1.19.4
Recreation steps : Start a new auth and proxy server and try to connect a new node as auth and proxy server.
Debug logs:
```
Proxy failed to establish connection to cluster: connection error: desc = "transport: Error while dialing failed to dial: dial tcp lb-ip:3025: i/o timeout". pid:11409.1 service/connect.go:119
2023-01-06T15:48:00Z ERRO [PROC:1] Node failed to establish connection to Teleport Proxy. We have tried the following endpoints: pid:11409.1 service/connect.go:1106
2023-01-06T15:48:00Z ERRO [PROC:1] - connecting to auth server directly: connection error: desc = "transport: Error while dialing failed to dial: dial tcp lb-ip:3025: i/o timeout" pid:11409.1 service/connect.go:1107
2023-01-06T15:48:00Z ERRO [PROC:1] - connecting to auth server through tunnel: connection error: desc = "transport: Error while dialing failed to dial: context deadline exceeded" pid:11409.1 service/connect.go:1111
2023-01-06T15:48:00Z ERRO [PROC:1] "Node failed to establish connection to cluster: Failed to connect to Auth Server directly or over tunnel, no methods remaining.\n\tconnection error: desc = "transport: Error while dialing failed to dial: dial tcp lb-ip:3025: i/o timeout", connection error: desc = "transport: Error while dialing failed to dial: context deadline exceeded"." pid:11409.1 service/connect.go:119
```

Load Balancer :
Listeners : 443 and 3025

Can connect to Web UI and register the nodes, cannot extend my env from a single auth/proxy node to a cluster.

Config File :

```
version: v2
teleport:
  auth_servers:
    - <cluster-name>:3025
  nodename: <nodename>
  log:
    output: stderr
    severity: INFO
  ca_pin: <ca_pin>
  join_params:
    token_name: <token>
    method: token
  storage:
    type: dynamodb
    region: <region>
    table_name: <tablename>
    audit_events_uri:  ['dynamodb://<tablename>', 'file:///var/lib/teleport/audit/events', 'stdout://']
    audit_sessions_uri: <s3_URL>
    audit_retention_period: 365d
proxy_service:
  enabled: true
  public_addr: <cluster_name>:443
  https_keypairs:
  - key_file: <pem file>
    cert_file: <pem file>
auth_service:
  enabled: true
  cluster_name: <cluster_name>
  license_file: <path to license file>
```

QUESTION : Should 3022 be a listener too?
The auth/proxy nodes are in different AZs.

**Comment by webvictim** [comment]
`i/o timeout` means simply that the proxy is unable to connect to the auth server address provided on port 3025. This is usually due to a firewall rule blocking the traffic or a lack of routing between the IPs. Make sure that the security groups/firewalls on the LB you're connecting to allow incoming traffic from the IP range being used by your proxy VPC.

Did you remove the auth LB address from the config you shared? The value is blank.

If you're running the `auth_service` and `proxy_service` in the same process, you shouldn't need to specify `teleport.auth_servers` in your config file. If you intend to run the proxy service separately, do this:

```
auth_service:
  enabled: false
```

You shouldn't need any listeners on port 3022.

**Reply by Ruchira-R** [reply]
Hiya,
   Yes, I want to run two more Auth-Proxy Nodes( Want to run both Auth and Proxy in the same node) in the same region but in different Availability Zones. Have verified that the Security Group and Firewall are configured properly (3025 and 443). If I'm using TLS Routing, I wouldn't need to open port 3025 now, do I? 
   Yes, the LB IP is redacted.
After I start running the first Auth-Proxy Node, what steps do I need to follow more Auth-Proxy Nodes to add to my cluster?

The steps that I'm currently following are :

-  Generate CA_PIN from the first Auth-Proxy Node
-  Generate INVITE_TOKEN with auth,node and proxy as the role
-  In the new node, where Auth-Proxy needs to be configured, I configure the ENV Vars (CA_PIN and INVITE_TOKEN)
-  I also configure the ADDR as (domain-name:3025, not working  || lb-ip:3025, not working || node-private-ip:3025, not working)

Is there any documentation on how I can add more nodes to make it HA.
Not sufficient, no clear cut steps : https://goteleport.com/docs/ver/11.x/deploy-a-cluster/high-availability/#configuration

All I need is steps on how I can make my cluster HA, I've been trying to find out how for a long time, the documentation is really cluttered, no way to figure out what/how to fix errors

**Reply by webvictim** [reply]
You can **remove** the `auth_servers` line from the config file if you're trying to add another auth/proxy in the same cluster, as it'll pick up the HA state from DynamoDB and make itself part of the cluster that way (and the load balancer will direct the traffic)

This method is detailed here: https://goteleport.com/docs/reference/backends/#dynamodb 

I appreciate that this doc does in fact say you need to set `auth_servers`, but this only applies when running separate auth and proxy instances. I will edit the doc to make this a little clearer.

**Reply by Ruchira-R** [reply]
Thanks @webvictim for the quick reply.

The setup is not working as expected, if I bring down the first Auth-Proxy, I cannot connect to any of my nodes, the error throw is 
![image](https://user-images.githubusercontent.com/44061338/213096179-d86b9c01-e770-4a4f-9b18-582ca39062e3.png)

I can reach the Web UI, can login, the only thing left is to connect to my nodes, which shows the error seen above.

Warning seen on the Second Auth-Proxy Node,

![image](https://user-images.githubusercontent.com/44061338/213113089-f9e3db04-f94f-4cba-8e89-35db8c9f6537.png)

Now, it seems to me, that the second proxy cannot connect to the node already registered, because of a lack of reverse tunneling.
Is that correct? What more do I need to add in my teleport.yaml file to fix this? (Ports 443 and 3025 are the listeners in the LB) 

The Nodes were registered to the first Auth-Proxy prior to conversion of my cluster to HA. (Although, the backend storage was Dynamo DB then as well.). Will that affect my ssh connections to the nodes, when the second and third Auth-Proxy nodes are brought up?

I'm missing tunnel_listen_addr in the First Auth-Proxy Node in proxy_service. Is this the reason that it isn't working?
 
It also seems to be related to this issue #19720

**Reply by webvictim** [reply]
1) If both of your auth/proxy nodes are registered with the same DynamoDB backend and are added to the same target group behind the load balancer (which exposes port 443 externally) then any nodes connected to your cluster should automatically spawn a second outbound reverse tunnel connection when the second proxy comes online.

It looks like you're providing your own TLS key and cert via the `https_keypairs` section of the config file - and as this is the case, I presume you want to multiplex all Teleport traffic over port 443. As such, you will need to enable multiplexing in the config file on all your auth/proxy instances:

```yaml
auth_service:
  proxy_listener_mode: multiplex
```

You should restart Teleport on all your auth/proxy services and then all your nodes after making this change. Your nodes should reconnect and establish their reverse tunnel connections correctly over port 443.

2) You can also remove the port 3025 listener from the load balancer if you're running auth/proxy services together, as nothing will need to make a connection directly to the auth service.
